File: framework/lib/radio_button.dart
part of uxml;

/**
 * Implements a button control that maintains a mutually exclusive checked
 * state among a group.
 *
 * @author:ferhat@ (Ferhat Buyukkokten)
 */
class RadioButton extends Button {

  static ElementDef radiobuttonElementDef;
  /** IsChecked property definition */
  static PropertyDefinition isCheckedProperty;
  /**
   * Group property definition. The group name is used as a key to
   * application level dictionary that keeps track of which radio button is
   * currently active among group members.
   */
  static PropertyDefinition groupProperty;
  static Map<String, RadioButton> _activeGroups;

  RadioButton() : super() {
  }

  /**
   * Sets/returns checked state.
   */
  bool get isChecked {
    return getProperty(isCheckedProperty);
  }

  set isChecked(bool value) {
    return setProperty(isCheckedProperty, value);
  }

  /**
   * Sets/returns radio button group name.
   */
  String get group {
    return getProperty(groupProperty);
  }

  set group(String name) {
    setProperty(groupProperty, name);
  }

  /** Overrides Button.onClick. */
  void onClick(EventArgs e) {
    bool prevChecked = isChecked;
    isChecked = true;
    EventArgs clickArgs = new EventArgs(this);
    clickArgs.event = Button.clickEvent;
    clickArgs.source = this;
    super.notifyListeners(Button.clickEvent, clickArgs);
    e.handled = true;
  }

  static void _isCheckedChangeHandler(UIElement target,
                                      PropertyDefinition property,
                                      Object oldValue,
                                      Object newValue) {
    RadioButton button = target;
    button._updateSharedGroupState(newValue);
  }

  static void _groupChangeHandler(UIElement target, PropertyDefinition property,
      Object oldValue, Object newValue) {
    RadioButton button = target;
    button._updateSharedGroupState(button.isChecked);
  }

  void _updateSharedGroupState(bool checked) {
    if (checked) {
      if (_activeGroups.containsKey(group)) {
        RadioButton activeButton = _activeGroups[group];
        if (activeButton != null && (activeButton != this)) {
          activeButton.setProperty(isCheckedProperty, false);
        }
      }
      _activeGroups[group] = this;
    }
  }

  /** @see UxmlElement.getDefinition. */
  ElementDef getDefinition() => radiobuttonElementDef;

  /** Registers component. */
  static void registerRadioButton() {
    _activeGroups = new Map<String, RadioButton>();
    isCheckedProperty = ElementRegistry.registerProperty("isChecked",
        PropertyType.BOOL, PropertyFlags.NONE, _isCheckedChangeHandler, false);
    groupProperty = ElementRegistry.registerProperty("group",
        PropertyType.STRING, PropertyFlags.NONE, _groupChangeHandler,
        "default");
    radiobuttonElementDef = ElementRegistry.register("RadioButton",
        Button.buttonElementDef,
        [isCheckedProperty, groupProperty],
        null);
  }
}

File: plugin/com.hello.uxml.tools.ui/src/com/hello/uxml/tools/framework/RadioButton.java
package com.hello.uxml.tools.framework;

import com.hello.uxml.tools.framework.events.MouseEventArgs;

import java.util.EnumSet;

/**
 * Implements a button control that maintains mutually exclusive isChecked state
 * among a group.
 *
 * @author ferhat
 */
public class RadioButton extends Button {

  /** IsChecked property definition */
  public static PropertyDefinition isCheckedPropDef = PropertySystem.register("IsChecked",
      Boolean.class, RadioButton.class,
      new PropertyData(false, EnumSet.of(PropertyFlags.None), new PropertyChangeListener() {
        @Override
        public void propertyChanged(PropertyChangedEvent e) {
          RadioButton button = (RadioButton) e.getSource();
          boolean checkValue = (Boolean) e.getNewValue();
          if (checkValue) {
            // We need to uncheck the active radio button in our group
            UIElement activeButton = Application.getCurrent().getRadioButton(button.getGroup());
            if (activeButton != null) {
              activeButton.setProperty(isCheckedPropDef, false);
            }
            Application.getCurrent().setRadioButton(button.getGroup(), button);
          }
        }}));

  /** Group property definition */
  public static PropertyDefinition groupPropDef = PropertySystem.register("Group",
      String.class, RadioButton.class,
      new PropertyData("", EnumSet.of(PropertyFlags.None)));

  /**
   * Sets/returns radio button group key.
   */
  public String getGroup() {
    return (String) getProperty(groupPropDef);
  }

  public void setGroup(String groupName) {
    setProperty(groupPropDef, groupName);
  }

  @Override public void close() {
    // unregister from button app level radiogroups
    if (getIsChecked()) {
      Application.getCurrent().setRadioButton(getGroup(), null);
    }
    super.close();
  }

  /**
   * Returns whether button is in checked state.
   */
  public boolean getIsChecked() {
    return (Boolean) getProperty(isCheckedPropDef);
  }

  /**
   * Sets button checked state.
   */
  public void setIsChecked(boolean value) {
    setProperty(isCheckedPropDef, value);
  }

  @Override
  protected void onMouseClick(MouseEventArgs e) {
    setIsChecked(true);
    super.onMouseClick(e);
  }
}
